[inductor] simplify expr when looking up size hint

ColinPeppler opened the pull request:
## Context

Suppose we have two symbols: `u0` and `s0` where we know that `u0 = s0`. Now, let's say we tried to look up the size hint for `u0 + 1`. 
* Before this PR, we would use a fallback hint if one was provided.
https://github.com/pytorch/pytorch/blob/3f6acf65fd9b6094513cf28898a42b90dd1169a0/torch/_inductor/sizevars.py#L406-L407

* With this PR, we would try to replace `u0` with `s0` via `simplify()` before using a fallback hint. https://github.com/pytorch/pytorch/blob/3f6acf65fd9b6094513cf28898a42b90dd1169a0/torch/_inductor/sizevars.py#L46-L47

## Concrete Example
A scenario where this is useful is when we're running autotuning benchmarking on bmm with two input nodes: one who has `s0` as the batch size and one who has `u0` as the batch size. During benchmarking, we'll create two example input tensors where the input with `u0` has to use a fallback hint for batch size. This will lead to a mismatch.

https://github.com/pytorch/pytorch/blob/e3d80f2fa98d7ab02f88023d381b2e5981dd99ff/torch/_inductor/select_algorithm.py#L991-L997

Using the fallback hint (i.e. 8192) leads to a batch size mismatch.

```
# Note: s0 = 7 and u0 = 7 and fallback hint is 8192.
LoweringException: ErrorFromChoice: Expected size for first two dimensions of batch2 tensor to be: [7, 30] but got: [8192, 30].
From choice ExternKernelCaller(extern_kernels.bmm)
```


Differential Revision: D55619331




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @amjames @desertfire @chauhang

ColinPeppler commented:
Hi @peterbell10, should this change be okay? or does this look like a fix that should be happening in Dynamo?

aakhundov commented on test/inductor/test_unbacked_symints.py:
```diff
@@ -156,6 +156,37 @@ def fn(x):
+            unbacked = x.expand(u0, *x.shape)
+            backed = x.expand(s0, *x.shape)
+
+            cat = torch.cat([backed, unbacked, unbacked], dim=1)  # [s0, 30, 16]
```
Could you add a comment above this line mentioning that here `s0` and `u0` are unified?

aakhundov reviewed (approved):
Thanks for the fix!

aakhundov commented:
> or does this look like a fix that should be happening in Dynamo?

I think, it's fine to do the replacements inside `size_hint`? cc @ezyang.

ColinPeppler commented on test/inductor/test_unbacked_symints.py:
```diff
@@ -156,6 +156,37 @@ def fn(x):
+            cat = torch.cat([backed, unbacked, unbacked], dim=1)  # [s0, 30, 16]
+            mat1 = torch.permute(cat, [0, 2, 1])  # [s0, 16, 30]
+            mat2 = w.expand(u0, *w.shape)  # [u0, 30, 32]
+            bmm = torch.ops.aten.bmm(mat1, mat2)  # [s0, 32, 7]
```
woops forgot to update lol

ColinPeppler committed:
[inductor] simplify expr when looking up size hint (#123140)

Summary:
## Context

Suppose we have two symbols: `u0` and `s0` where we know that `u0 = s0`. Now, let's say we tried to look up the size hint for `u0 + 1`. 
* Before this PR, we would use a fallback hint if one was provided.
https://github.com/pytorch/pytorch/blob/3f6acf65fd9b6094513cf28898a42b90dd1169a0/torch/_inductor/sizevars.py#L406-L407

* With this PR, we would try to replace `u0` with `s0` via `simplify()` before using a fallback hint. https://github.com/pytorch/pytorch/blob/3f6acf65fd9b6094513cf28898a42b90dd1169a0/torch/_inductor/sizevars.py#L46-L47

## Concrete Example
A scenario where this is useful is when we're running autotuning benchmarking on bmm with two input nodes: one who has `s0` as the batch size and one who has `u0` as the batch size. During benchmarking, we'll create two example input tensors where the input with `u0` has to use a fallback hint for batch size. This will lead to a mismatch.

https://github.com/pytorch/pytorch/blob/e3d80f2fa98d7ab02f88023d381b2e5981dd99ff/torch/_inductor/select_algorithm.py#L991-L997

Using the fallback hint (i.e. 8192) leads to a batch size mismatch.


```python
# Note: s0 = 7 and u0 = 7 and fallback hint is 8192.
LoweringException: ErrorFromChoice: Expected size for first two dimensions of batch2 tensor to be: [7, 30] but got: [8192, 30].
From choice ExternKernelCaller(extern_kernels.bmm)
```




Test Plan:
CI

```
$ CUDA_VISIBLE_DEVICES=0 python test/inductor/test_unbacked_symints.py -k test_equivalent_backed_unbacked_cuda

### Before ###
  File "torch/_inductor/select_algorithm.py", line 964, in __call__
    timings = do_autotuning(precompile_fn)
  File "torch/_inductor/select_algorithm.py", line 911, in do_autotuning
    timings = self.lookup(
  File "torch/_inductor/codecache.py", line 306, in lookup
    raise e
  File "torch/_inductor/codecache.py", line 297, in lookup
    timings = benchmark(choices)
  File "torch/_inductor/select_algorithm.py", line 897, in autotune
    return make_benchmark_fn()(choices)
  File "torch/_inductor/select_algorithm.py", line 1068, in benchmark_in_current_process
    raise ErrorFromChoice(msg, choice, debug_str())  # noqa: TRY200

torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: ErrorFromChoice: Expected size for first two dimensions of batch2 tensor to be: [7, 30] but got: [8192, 30].

From choice ExternKernelCaller(extern_kernels.bmm)
inputs = [
    torch.empty_strided((7, 30, 16), (480, 16, 1), dtype=torch.float32, device='cuda'),
    torch.empty_strided((30, 32), (32, 1), dtype=torch.float32, device='cuda'),
]


### After ###
----------------------------------------------------------------------
Ran 1 test in 4.627s

OK
```

Reviewed By: tissue3, aakhundov

Differential Revision: D55619331

ColinPeppler commented:
CI passes in OSS and internally. Also, we were worried that `self.simplify(expr)` will fail if `expr` is an int but `sympy.expand(expr)` can handle if `expr` is an int.

ColinPeppler commented:
@pytorchbot merge

peterbell10 commented on torch/_inductor/sizevars.py:
```diff
@@ -415,6 +415,7 @@ def symbolic_hint(self, expr: Expr) -> Expr:
         return sympy_subs(expr, self.var_to_val)
 
     def size_hint(self, expr: Expr, *, fallback: Optional[int] = None) -> int:
+        expr = self.simplify(expr)
```
This should be in `symbolic_hint` but otherwise this LGTM

peterbell10 replied:
@ColinPeppler this wasn't fixed

clee2000 commented:
@pytorchbot merge -f "merged internally"
